### PR TITLE
Added TensorFlow support to `nncf.Tensor`

### PR DESCRIPTION
**Pull request by olegkkruglov**
### Changes

* Added `tf_numeric.py` and `tf_linalg.py` files with implementations of methods needed for `nncf.Tensor` support.
* Added TensorFlow backend for `nncf.Tensor`.
* Fixed bug in `__ifloordiv__` operator for `nncf.Tensor`.
* Added related tests.

### Reason for changes

Currently TensorFlow tensors are not supported by `nncf.Tensor`. It prevents https://github.com/openvinotoolkit/nncf/issues/3041 from being done.

### Related tickets

https://github.com/openvinotoolkit/nncf/issues/3041

### Tests

`TestTFNNCFTensorOperators` and `TestGPUTFNNCFTensorOperators` classes were added to `tests/tensorflow/test_tensor.py`. Some changes were necessary for `tests/cross_fw/test_templates/template_test_nncf_tensor.py`, mostly related to different device management in TensorFlow.

### CODE REVIEW

**Comment by alexsu52** [comment]
Please, update your branch.

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py [hunk 1]
```diff
@@ -112,7 +112,8 @@ def test_operators_tensor(self, op_name):
         assert res.dtype == res_nncf.data.dtype
         assert all(res == res_nncf.data)
         assert isinstance(res_nncf, Tensor)
-        assert res_nncf.device == nncf_tensor_a.device
+        if not (self.backend() == TensorBackend.tf and self.device() == TensorDeviceType.CPU):
```
Comment (kshpv) [inline tests/cross_fw/test_templates/template_test_nncf_tensor.py:115]: Could you briefly explain the reason for these changes while you return specific device tensors for TF?

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py [hunk 2]
```diff
@@ -158,7 +164,7 @@ def test_comparison_tensor(self, op_name):
         res = fn(tensor_a, tensor_b)
         res_nncf = fn(nncf_tensor_a, nncf_tensor_b)
 
-        assert res == res_nncf
+        assert res_nncf == res
```
Comment (kshpv) [inline tests/cross_fw/test_templates/template_test_nncf_tensor.py:167]: Could you clarify why you changed the order here?

File: nncf/tensor/functions/tf_numeric.py [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+
+
+@numeric.as_tensor_like.register(tf.Tensor)
+def _(a: tf.Tensor, data: Any) -> tf.Tensor:
+    with tf.device(a.device):
+        return tf.convert_to_tensor(data)
+
+
+@numeric.item.register(tf.Tensor)
+def _(a: tf.Tensor) -> Union[int, float, bool]:
```
[the remark this answers is not in the document]
Comment (olegkkruglov) [reply nncf/tensor/functions/tf_numeric.py:313]: it still returns a tensor (with empty shape) and the following logic is still necessary. but the addition of reshape here allows to avoid deprecation warning, so, I'll add it. thanks for pointing out.

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py (hunk 2 above)
Comment (olegkkruglov) [reply tests/cross_fw/test_templates/template_test_nncf_tensor.py:167]: with the old order comparison operator for `tf.Tensor` was called here, it tries to convert `nncf.Tensor` to `tf.EagerTensor`. It causes a problem because it calls `__len__` operator somewhere inside. I tried to implement `__len__` operator for `nncf.Tensor` as `return len(self._data)` but the comparison still fails for scalar tensors. currently I have no idea how to resolve it and used changing order here as a quick workaround. I'd appreciate any other suggestions.

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py (hunk 1 above)
Comment (olegkkruglov) [reply tests/cross_fw/test_templates/template_test_nncf_tensor.py:115]: the result of binary operation with two CPU TF tensors is GPU TF tensor by default. it means that the following check would fail on CPU.

**Review by alexsu52 — CHANGES_REQUESTED** [review]
Thanks for contributing. Please address my comments. 
PS: Sorry for delay with reviewing.

File: nncf/tensor/functions/tf_numeric.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+@numeric.concatenate.register(tf.Tensor)
+def _(x: List[tf.Tensor], axis: int = 0) -> tf.Tensor:
+    with tf.device(x[0].device):
+        return tf.concat(x, axis=axis)
+
+
+@numeric.unstack.register(tf.Tensor)
+def _(x: tf.Tensor, axis: int = 0) -> List[tf.Tensor]:
+    with tf.device(x.device):
+        if not list(x.shape):
```
[the remark this answers is not in the document]
Comment (olegkkruglov) [reply nncf/tensor/functions/tf_numeric.py:199]: I just tried to do the same behavior as for torch. Is it unnecessary here?
https://github.com/openvinotoolkit/nncf/blob/909ce0ac03f78b890d9fb774da153aaaf9b59e1a/nncf/tensor/functions/torch_numeric.py#L195

File: nncf/tensor/functions/tf_numeric.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+    with tf.device(a.device):
+        return tf.reshape(a, [-1])
+
+
+@numeric.max.register(tf.Tensor)
+def _(a: tf.Tensor, axis: Optional[Union[int, Tuple[int, ...]]] = None, keepdim: bool = False) -> tf.Tensor:
+    with tf.device(a.device):
+        if axis is None:
+            return tf.reduce_max(a)
+        return tf.reduce_max(a, axis=axis, keepdims=keepdim)
```
[the remark this answers is not in the document]
Comment (olegkkruglov) [reply nncf/tensor/functions/tf_numeric.py]: I've recognized that torch part have the same issue (`keepdims` is always `False` in case if `axis` is `None`). It works because `torch.allclose` function does not compare shapes (as well as in TensorFlow). Was it intentional? And should I fix torch part here as well if it was not?

File: nncf/tensor/functions/tf_linalg.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    with tf.device(a.device):
+        if driver is not None:
+            warnings.warn("Driver specifying is not supported in TensorFlow lstsq method")
+        if tf.rank(b) == 1:
+            b = tf.expand_dims(b, axis=0)
+        perm = list(range(tf.rank(b)))
+        perm[-1], perm[-2] = perm[-2], perm[-1]
+        b = tf.transpose(b, perm=perm)
+
+        return tf.linalg.lstsq(a, b)
```
[the remark this answers is not in the document]
Comment (olegkkruglov) [reply nncf/tensor/functions/tf_linalg.py:121]: Existing testcases are failing without the transpose. Could you please clarify what tests do you want to add?

**Comment by kshpv** [comment]
It looks like the PR is stuck. I see that @olegkkruglov answered all the comments from @alexsu52 
@alexsu52 did you have a chance to take a look at the updated PR?

**Review by kshpv — APPROVED** [review]

**Review by alexsu52 — COMMENTED** [review]
Please, rebase the PR

File: nncf/tensor/functions/tf_numeric.py [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+
+
+@numeric.median.register(tf.Tensor)
+def _(
+    a: tf.Tensor,
+    axis: Union[int, Tuple[int, ...]] = None,
+    keepdims: bool = False,
+) -> tf.Tensor:
+    numpy_a = np.array(a)
+    numpy_median = np.median(numpy_a, axis=axis, keepdims=keepdims)
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/tensor/functions/tf_numeric.py]: Do you have any evidence that this implementation is more performant than calling `np.median`?

File: nncf/tensor/functions/tf_numeric.py [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+        median = tf.squeeze(median, axis=-1)
+        if keepdims:
+            for axe in sorted(axis, key=lambda x: abs(x)):
+                median = tf.expand_dims(median, axe)
+
+        return median
+
+
+@numeric.round.register(tf.Tensor)
+def _(a: tf.Tensor, decimals=0) -> tf.Tensor:
```
Comment (alexsu52) [inline nncf/tensor/functions/tf_numeric.py:254]:
```suggestion
def _(a: tf.Tensor, decimals: int = 0) -> tf.Tensor:
```

**Review by alexsu52 — COMMENTED** [review]
Looks like the pull request is in the home stretch. Please address my comments and add support for `as_numpy_tensor` function from develop as well.

**Comment by darshil929** [comment]
Hello! @alexsu52 @kshpv

Could you please quickly brief me on the current progress of the PR, which comments still need to be addressed, and what would be the best starting point for me to take over? Also, any tips on how to effectively branch from the PR would be super helpful. 

Thanks!

**Comment by alexsu52** [comment]
> Hello! @alexsu52 @kshpv
> 
> I've been assigned to work on issue #[3041](https://github.com/openvinotoolkit/nncf/issues/3041) and I'd like to pick up where PR #3106 was left off. Could you please quickly brief me on the current progress of the PR, which comments still need to be addressed, and what would be the best starting point for me to take over? Also, any tips on how to effectively branch from the PR would be super helpful.
> 
> Thanks!

Hello @darshil929,

I would suggest to start from creating the branch in your fork, rebase it on develop and open the PR where you will work on resolving all unresolved comments.

How checking out pull requests: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally

**Comment by MaximProshin** [comment]
@alexsu52 , can this PR be closed?
